Fix a bug finding the quantum function jaxpr when using quantum primitives with dynamic one-shot

mehrdad2m opened the pull request:
**Context:**
When dealing with lowering of quantum operations, we need to extract the quantum function jaxpr by finding the function primitive in the circuit jaxpr which is assumed to be the first element in jaxpr.eqns. However, this assumption is wrong specifically when dynamic-one-shot is used. In this case the quantum function is wrapped in another funcion, therefore we get a crash trying to access the func primitive at jaxpr.eqns[0]. 

**Description of the Change:**
implemented a function def _get_call_jaxpr(jaxpr) which iterates through equations in the jaxpr and finds the func primitive. 

**Benefits:**
No specific assumption is made on the location of the quantum function in the jaxpr

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-68756]

paul0403 commented:
Nice catch :100:! Just wondering what syndrome this is fixing, is there an example (issue/story) of the current failure?

paul0403 commented on frontend/catalyst/jax_primitives.py:
```diff
@@ -662,6 +662,15 @@ def _grad_abstract(*args, jaxpr, fn, grad_params):
+    for eqn in jaxpr.eqns:
+        primitive = eqn.primitive
+        if primitive is func_p:
+            return eqn.params.get("call_jaxpr")
```
I think `dict.get("key")` will silently return None when the key is not present vs `dict["key"]` will actively bark a KeyError. I know it isn't really likely to happen but if jax decides to remove the `"call_jaxpr"` key of the func primitive then we should actively fail here.

mehrdad2m commented:
> **Context:**
> 
> **Description of the Change:**
> 
> **Benefits:**
> 
> **Possible Drawbacks:**
> 
> **Related GitHub Issues:**



> **Context:**
> 
> **Description of the Change:**
> 
> **Benefits:**
> 
> **Possible Drawbacks:**
> 
> **Related GitHub Issues:**



> Nice catch 💯! Just wondering what syndrome this is fixing, is there an example (issue/story) of the current failure?



> Nice catch 💯! Just wondering what syndrome this is fixing, is there an example (issue/story) of the current failure?

Hi @paul0403, Thanks for the review. I have updated the PR description and added a test.

dime10 commented on frontend/catalyst/jax_primitives.py:
```diff
@@ -684,8 +693,8 @@ def _grad_lowering(ctx, *args, jaxpr, fn, grad_params):
     diffArgIndices = ir.DenseIntElementsAttr.get(argnum_numpy)
-
-    _func_lowering(ctx, *args, call_jaxpr=jaxpr.eqns[0].params["call_jaxpr"], fn=fn, call=False)
+    func_call_jaxpr = _get_call_jaxpr(jaxpr)
+    _func_lowering(ctx, *args, call_jaxpr=func_call_jaxpr, fn=fn, call=False)
```
Was this also broken for the grad lowering or did you just to update it pro-actively? :)

dime10 replied:
Ah good to know. Could you add those tests and mark them xfail (or skip if they crash python)? :)

mehrdad2m replied:
I added some tests that are failing right now.

dime10 commented on frontend/test/pytest/test_mid_circuit_measurement.py:
```diff
@@ -391,6 +393,37 @@ def circuit(x):
+
+        assert np.allclose(expected, observed)
+
+    def test_get_call_jaxpr(self):
+        """Test _get_call_jaxpr raises AsserionError if no function primitive exists."""
+
+        def f(x):
+            return x * x
+
+        jaxpr = make_jaxpr(f)(2.0)
+        with pytest.raises(AssertionError, match="No call_jaxpr found in the JAXPR"):
+            _ = _get_call_jaxpr(jaxpr)
```
Since this test is not related to mcms, could you move it `pytest/test_jax_primitives.py` (creating it if it doesn't exist) ?

dime10 commented on frontend/test/pytest/test_mid_circuit_measurement.py:
```diff
@@ -391,6 +393,37 @@ def circuit(x):
+            g = qml.QNode(circuit, dev, mcm_method=mcm_method)
+            return mitigate_with_zne(g, scale_factors=s)()
+
+        observed = mitigated_circuit()
+        mcm_method = None
+        expected = mitigated_circuit()
+
+        assert np.allclose(expected, observed)
```
Aren't we testing it against itself here? 🤔

paul0403 replied:
I think python late binding means reference to globals within functions will actually track the latest value of a global symbol, although not entirely sure here. In any case I agree that the test should make two qnodes explicitly, one for each `mcm_method`, and compare them.

dime10 replied:
Oh I see, yeah I missed that.

dime10 replied:
But in this case we are not invoking the python function twice, only the compiled function, which is not going to see any changes in the `mcm_method` variable.

paul0403 replied:
Ah true, the second call doesn't go through python at all. Then it should be two qnodes.

mehrdad2m replied:
Ah good catch! I forgot I am calling compiled funnction :) Fixed it.

dime10 commented on frontend/catalyst/jax_primitives.py:
```diff
@@ -662,6 +662,15 @@ def _grad_abstract(*args, jaxpr, fn, grad_params):
     return tuple(transformed_signature.get_results())
 
 
+def _get_call_jaxpr(jaxpr):
+    """Extracts the `call_jaxpr` from a JAXPR if it exists.""" ""
+    for eqn in jaxpr.eqns:
+        primitive = eqn.primitive
+        if primitive is func_p:
+            return eqn.params["call_jaxpr"]
+    raise AssertionError("No call_jaxpr found in the JAXPR.")
```
Can't believe the fix was so simple 😅 

Can you confirm that the design matches what we expect? I think what would need to happen is
- the folding is applied to the quantum function generated inside the dynamic one shot function
- the fitting is applied to the entire dynamic one shot function, which includes the execution loop and post-processing

mehrdad2m replied:
I am not too familiar with the ZNE algorithm, but as far as I understand now, this is what happens. I can see that the extrapolation is happening at a higher level on the results of the zne_p primitive which uses the dynamic-one-shot function. I am not too sure where the folding happens but I can see that the quantum function (not the dynamic one-shot function) is used to create the lowered ZNE operation which is where I assume the folding happens. I can dig deeper more next week though :)

dime10 replied:
Nice, just wanted to confirm. Good work :)

dime10 commented on frontend/catalyst/api_extensions/differentiation.py:
```diff
@@ -801,8 +801,6 @@ def _make_jaxpr_check_differentiable(
     with mark_gradient_tracing(method):
         jaxpr, shape = jax.make_jaxpr(f, return_shape=True)(*args, **kwargs)
     _, out_tree = tree_flatten(shape)
-    assert len(jaxpr.eqns) == 1, "Expected jaxpr consisting of a single function call."
-    assert jaxpr.eqns[0].primitive == func_p, "Expected jaxpr consisting of a single function call."
```
Probably similar to my other comment, is this addressing a second bug regarding derivatives and dynamic one shot, or is it related to ZNE somehow? :)

mehrdad2m replied:
no this is not related to ZNE, it happens when we use derivatives with dynamic one-shot

dime10 reviewed (commented):
Thank you @mehrdad2m! :)
